// access/lib/errors.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
var PropertyUndefinedException = /** @class */ (function (_super) {
    __extends(PropertyUndefinedException, _super);
    function PropertyUndefinedException(message) {
        var _this = _super.call(this, message) || this;
        _this.name = 'PropertyUndefinedException';
        return _this;
    }
    return PropertyUndefinedException;
}(Error));
exports.PropertyUndefinedException = PropertyUndefinedException;
var PropertyTypeMismatchException = /** @class */ (function (_super) {
    __extends(PropertyTypeMismatchException, _super);
    function PropertyTypeMismatchException(message) {
        var _this = _super.call(this, message) || this;
        _this.name = 'PropertyTypeMismatchException';
        return _this;
    }
    return PropertyTypeMismatchException;
}(Error));
exports.PropertyTypeMismatchException = PropertyTypeMismatchException;
var PropertyMissingException = /** @class */ (function (_super) {
    __extends(PropertyMissingException, _super);
    function PropertyMissingException(message) {
        var _this = _super.call(this, message) || this;
        _this.name = 'PropertyMissingException';
        return _this;
    }
    return PropertyMissingException;
}(Error));
exports.PropertyMissingException = PropertyMissingException;
function createPropertyUndefinedException(object, accessorFunction, fallback) {
    return new PropertyUndefinedException("Accessing object (" + JSON.stringify(object) + ") with accessor (" + accessorFunction.toString() + ") resulted in undefined. Fell back to " + fallback);
}
exports.createPropertyUndefinedException = createPropertyUndefinedException;
function createPropertyTypeMismatchException(expectedType, object, accessorFunction, fallback) {
    return new PropertyTypeMismatchException("Accessing object (" + JSON.stringify(object) + ") with accessor (" + accessorFunction.toString() + ") resulted in undefined. Expected type (" + expectedType + "). Fell back to " + fallback);
}
exports.createPropertyTypeMismatchException = createPropertyTypeMismatchException;
function createPropertyMissingException(object, accessorFunction, fallback) {
    return new PropertyMissingException("No property was found in object (" + JSON.stringify(object) + ") with accessor (" + accessorFunction.toString() + "). Fell back to " + fallback);
}
exports.createPropertyMissingException = createPropertyMissingException;
