// access/lib/getters.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var typeGuards_1 = require("./typeGuards");
var errors_1 = require("./errors");
function getWithGuard(guard, object, accessorFunction, fallback, fallbackCallback) {
    try {
        var result = accessorFunction(object);
        if (result === undefined)
            throw errors_1.createPropertyUndefinedException(object, accessorFunction, fallback);
        if (guard.check(fallback) && guard.check(result))
            return result;
        throw errors_1.createPropertyTypeMismatchException(guard.type, object, accessorFunction, fallback);
    }
    catch (error) {
        if (fallbackCallback && error instanceof TypeError) {
            fallbackCallback(errors_1.createPropertyMissingException(object, accessorFunction, fallback));
        }
        else if (fallbackCallback) {
            fallbackCallback(error);
        }
        return fallback;
    }
}
function getString(object, accessorFunction, fallback, fallbackCallback) {
    return getWithGuard(typeGuards_1.StringGuard, object, accessorFunction, fallback, fallbackCallback);
}
exports.getString = getString;
function getNumber(object, accessorFunction, fallback, fallbackCallback) {
    return getWithGuard(typeGuards_1.NumberGuard, object, accessorFunction, fallback, fallbackCallback);
}
exports.getNumber = getNumber;
function getBoolean(object, accessorFunction, fallback, fallbackCallback) {
    return getWithGuard(typeGuards_1.BooleanGuard, object, accessorFunction, fallback, fallbackCallback);
}
exports.getBoolean = getBoolean;
function getStringArray(object, accessorFunction, fallback, fallbackCallback) {
    return getWithGuard(typeGuards_1.StringArrayGuard, object, accessorFunction, fallback, fallbackCallback);
}
exports.getStringArray = getStringArray;
function getNumberArray(object, accessorFunction, fallback, fallbackCallback) {
    return getWithGuard(typeGuards_1.NumberArrayGuard, object, accessorFunction, fallback, fallbackCallback);
}
exports.getNumberArray = getNumberArray;
function getBooleanArray(object, accessorFunction, fallback, fallbackCallback) {
    return getWithGuard(typeGuards_1.BooleanArrayGuard, object, accessorFunction, fallback, fallbackCallback);
}
exports.getBooleanArray = getBooleanArray;
function getStringMap(object, accessorFunction, fallback, fallbackCallback) {
    return getWithGuard(typeGuards_1.StringMapGuard, object, accessorFunction, fallback, fallbackCallback);
}
exports.getStringMap = getStringMap;
function getNumberMap(object, accessorFunction, fallback, fallbackCallback) {
    return getWithGuard(typeGuards_1.NumberMapGuard, object, accessorFunction, fallback, fallbackCallback);
}
exports.getNumberMap = getNumberMap;
function getBooleanMap(object, accessorFunction, fallback, fallbackCallback) {
    return getWithGuard(typeGuards_1.BooleanMapGuard, object, accessorFunction, fallback, fallbackCallback);
}
exports.getBooleanMap = getBooleanMap;
