// access/lib/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var getters_1 = require("./getters");
exports.getString = getters_1.getString;
exports.getNumber = getters_1.getNumber;
exports.getBoolean = getters_1.getBoolean;
exports.getStringArray = getters_1.getStringArray;
exports.getNumberArray = getters_1.getNumberArray;
exports.getBooleanArray = getters_1.getBooleanArray;
exports.getStringMap = getters_1.getStringMap;
exports.getNumberMap = getters_1.getNumberMap;
exports.getBooleanMap = getters_1.getBooleanMap;
function accessFactory(fallbackCallback) {
    return {
        getString: function (object, accessorFunction, fallback) { return getters_1.getString(object, accessorFunction, fallback, fallbackCallback); },
        getNumber: function (object, accessorFunction, fallback) { return getters_1.getNumber(object, accessorFunction, fallback, fallbackCallback); },
        getBoolean: function (object, accessorFunction, fallback) { return getters_1.getBoolean(object, accessorFunction, fallback, fallbackCallback); },
        getStringArray: function (object, accessorFunction, fallback) { return getters_1.getStringArray(object, accessorFunction, fallback, fallbackCallback); },
        getNumberArray: function (object, accessorFunction, fallback) { return getters_1.getNumberArray(object, accessorFunction, fallback, fallbackCallback); },
        getBooleanArray: function (object, accessorFunction, fallback) { return getters_1.getBooleanArray(object, accessorFunction, fallback, fallbackCallback); },
        getStringMap: function (object, accessorFunction, fallback) { return getters_1.getStringMap(object, accessorFunction, fallback, fallbackCallback); },
        getNumberMap: function (object, accessorFunction, fallback) { return getters_1.getNumberMap(object, accessorFunction, fallback, fallbackCallback); },
        getBooleanMap: function (object, accessorFunction, fallback) { return getters_1.getBooleanMap(object, accessorFunction, fallback, fallbackCallback); },
    };
}
exports.default = accessFactory;
