// access/lib/typeGuards.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.StringGuard = {
    type: 'string',
    check: isString,
};
exports.NumberGuard = {
    type: 'number',
    check: isNumber,
};
exports.BooleanGuard = {
    type: 'boolean',
    check: isBoolean,
};
exports.StringArrayGuard = {
    type: 'stringArray',
    check: isStringArray,
};
exports.NumberArrayGuard = {
    type: 'numberArray',
    check: isNumberArray,
};
exports.BooleanArrayGuard = {
    type: 'booleanArray',
    check: isBooleanArray,
};
exports.StringMapGuard = {
    type: 'stringMap',
    check: isStringMap,
};
exports.NumberMapGuard = {
    type: 'numberMap',
    check: isNumberMap,
};
exports.BooleanMapGuard = {
    type: 'booleanMap',
    check: isBooleanMap,
};
function isString(obj) {
    return typeof obj === 'string';
}
function isNumber(obj) {
    return typeof obj === 'number';
}
function isBoolean(obj) {
    return typeof obj === 'boolean';
}
function isArray(obj) {
    return Array.isArray(obj);
}
function isObject(obj) {
    return typeof obj === 'object' && obj !== null;
}
function isStringArray(obj) {
    return isArray(obj) && obj.every(isString);
}
function isNumberArray(obj) {
    return isArray(obj) && obj.every(isNumber);
}
function isBooleanArray(obj) {
    return isArray(obj) && obj.every(isBoolean);
}
function isStringMap(obj) {
    return (isObject(obj) &&
        Object.keys(obj)
            .map(function (key) { return obj[key]; })
            .every(isString));
}
function isNumberMap(obj) {
    return (isObject(obj) &&
        Object.keys(obj)
            .map(function (key) { return obj[key]; })
            .every(isNumber));
}
function isBooleanMap(obj) {
    return (isObject(obj) &&
        Object.keys(obj)
            .map(function (key) { return obj[key]; })
            .every(isBoolean));
}
